src/typeEvaluator/typeEvaluate.ts

```diff
@@ -145,44 +230,29 @@ function handleObjectNode(node: ObjectNode, scope: Scope) {
     }
 
     if (attr.type === 'ObjectSplat') {
-      const value = walk({node: attr.value, scope})
-      $trace('object.splat.value %O', value)
-      mapObjectSplat(value, scope, (node) => {
-        for (const name in node.attributes) {
-          if (!Object.hasOwn(node.attributes, name)) {
-            continue
-          }
-
-          attributes[name] = node.attributes[name]
+      const result = handleObjectSplatNode(attr, scope)
+      $trace('object.splat.result %O', result)
+      if (result.type !== 'object') {
+        return result
+      }
+      for (const name in result.attributes) {
+        if (!result.attributes.hasOwnProperty(name)) {
+          continue
         }
-      })
+        attributes[name] = result.attributes[name]
+      }
     }
     if (attr.type === 'ObjectConditionalSplat') {
-      const condition = resolveCondition(attr.condition, scope)
-      $trace('object.conditional.splat.condition %O', condition)
-      if (condition || condition === undefined) {
-        const value = walk({node: attr.value, scope})
-
-        mapObjectSplat(value, scope, (node) => {
-          for (const name in node.attributes) {
-            if (!Object.hasOwn(node.attributes, name)) {
-              continue
-            }
-            const attribute = node.attributes[name]
-
-            if (condition) {
-              attributes[name] = attribute
-            } else if (condition === undefined) {
-              attributes[name] = {
-                type: 'objectAttribute',
-                value: attribute.value,
-                optional: true,
-              }
-            } else {
-              throw new Error('Unexpected condition')
-            }
-          }
-        })
+      const result = handleObjectConditionalSplatNode(attr, scope)
+      $trace('object.conditional.splat.result %O', result)
+      if (result.type !== 'object') {
+        return result
+      }
```

Comment:
I don't think this makes sense: If the splatted node is _not_ an object then we return that as the result of the whole object?

In the following query:

```groq
*[_type == "author"][0] {
    _type == "author" => select(age >= 18 => {lastname}, {firstname}),
    _id,
    _type
  }
```

Then we now end up generating:

```js
{
  "type": "union",
  "of": Array [
    Object {
      "type": "object",
      "attributes": Object {
        "firstname": Object {
          "type": "objectAttribute",
          "value": Object {
            "type": "string",
          },
        },
      },
    },
    Object {
      "type": "object",
      "attributes": Object {
        "lastname": Object {
          "type": "objectAttribute",
          "value": Object {
            "type": "string",
          },
        },
      },
    },
    Object {
      "type": "null",
    },
  ],
}
```

Because the "conditional splat" turns into a `union` and thus we hit this case.